File: dotnet/src/Obscura.Render/Core/Border.cs
// RECONCILIATION NOTE: this file ports `crates/obscura-render/src/border.rs`, which is owned
// by the border component agent. `LayoutStyle` has `BorderModel` and `OutlineModel` fields, so
// the types could not be deferred. The port is complete and faithful (including the border.rs
// unit tests, in RenderCoreTests.cs); the border agent should adopt this file rather than
// adding a second copy under a different folder.
using RgbaColor = Obscura.Render.Css.RgbaColor;

namespace Obscura.Render;

/// <summary>Four physical sides in CSS top-right-bottom-left order.</summary>
public readonly record struct Sides<T>(T Top, T Right, T Bottom, T Left)
{
    public static Sides<T> All(T value) => new(value, value, value, value);

    public Sides<TOut> Map<TOut>(Func<T, TOut> f) => new(f(Top), f(Right), f(Bottom), f(Left));

    public T[] AsArray() => [Top, Right, Bottom, Left];
}

/// <summary>Border and outline helpers shared by style, layout, and paint.</summary>
public static class BorderSides
{
    /// <summary>CSS's initial <c>medium</c> border and outline width in CSS pixels.</summary>
    public const float MediumBorderWidth = 3.0f;

    /// <summary>Expand a CSS 1-4 value list in top-right-bottom-left order.</summary>
    public static Sides<T>? ExpandSides<T>(ReadOnlySpan<T> values) => values.Length switch
    {
        1 => Sides<T>.All(values[0]),
        2 => new Sides<T>(values[0], values[1], values[0], values[1]),
        3 => new Sides<T>(values[0], values[1], values[2], values[1]),
        4 => new Sides<T>(values[0], values[1], values[2], values[3]),
        _ => null,
    };

    /// <summary>Expand a CSS 1-4 value list in top-right-bottom-left order.</summary>
    public static Sides<T>? ExpandSides<T>(IReadOnlyList<T> values)
    {
        T[] copy = new T[values.Count];
        for (int i = 0; i < values.Count; i++)
        {
            copy[i] = values[i];
        }

        return ExpandSides<T>(copy.AsSpan());
    }

    internal static float UsedWidth(float specified, BorderStyle style) =>
        style.IsVisible() ? F32.Max(specified, 0f) : 0f;
}

/// <summary>The CSS border-line styles represented by the renderer.</summary>
public enum BorderStyle
{
    /// <summary>The default.</summary>
    None = 0,

    Hidden,
    Dotted,
    Dashed,
    Solid,
    Double,
    Groove,
    Ridge,
    Inset,
    Outset,

    /// <summary><c>outline-style:auto</c>; border sides themselves never use this value.</summary>
    Auto,
}

public static class BorderStyleExtensions
{
    /// <summary>Whether this style contributes a used width and visible border paint.</summary>
    public static bool IsVisible(this BorderStyle style) =>
        style is not (BorderStyle.None or BorderStyle.Hidden);

    public static string CssName(this BorderStyle style) => style switch
    {
        BorderStyle.None => "none",
        BorderStyle.Hidden => "hidden",
        BorderStyle.Dotted => "dotted",
        BorderStyle.Dashed => "dashed",
        BorderStyle.Solid => "solid",
        BorderStyle.Double => "double",
        BorderStyle.Groove => "groove",
        BorderStyle.Ridge => "ridge",
        BorderStyle.Inset => "inset",
        BorderStyle.Outset => "outset",
        BorderStyle.Auto => "auto",
        _ => "none",
    };
}

/// <summary>One CSS <c>&lt;length-percentage&gt;</c> component of a corner radius.</summary>
public readonly record struct RadiusValue(float Length, float Percentage)
{
    public static RadiusValue Pixels(float length) => new(length, 0f);

    /// <summary>Fraction of the corresponding border-box axis (<c>0.5</c> is <c>50%</c>).</summary>
    public static RadiusValue FromPercentage(float percentage) => new(0f, percentage);

    public float Resolve(float axis) => F32.Max(Length + (Percentage * axis), 0f);

    public bool IsZero() => Length == 0f && Percentage == 0f;
}

/// <summary>Horizontal and vertical radii for one corner.</summary>
public readonly record struct CornerRadius(RadiusValue X, RadiusValue Y)
{
    public static CornerRadius Circular(RadiusValue value) => new(value, value);
}

/// <summary>Four elliptical radii in clockwise order from the top-left corner.</summary>
public readonly record struct BorderRadii(
    CornerRadius TopLeft,
    CornerRadius TopRight,
    CornerRadius BottomRight,
    CornerRadius BottomLeft)
{
    public static BorderRadii All(CornerRadius radius) => new(radius, radius, radius, radius);

    /// <summary>
    /// Resolve percentages and apply CSS Backgrounds 3's corner-overlap rule. One common scale
    /// factor preserves every ellipse's aspect ratio.
    /// </summary>
    public ResolvedBorderRadii Resolve(float width, float height)
    {
        ResolvedBorderRadii resolved = new(
            ResolveCorner(TopLeft, width, height),
            ResolveCorner(TopRight, width, height),
            ResolveCorner(BottomRight, width, height),
            ResolveCorner(BottomLeft, width, height));

        static float Ratio(float available, float requested) =>
            requested > 0f ? F32.Max(available, 0f) / requested : 1f;

        float scale = F32.Min(
            F32.Min(
                F32.Min(
                    F32.Min(1f, Ratio(width, resolved.TopLeft.X + resolved.TopRight.X)),
                    Ratio(width, resolved.BottomLeft.X + resolved.BottomRight.X)),
                Ratio(height, resolved.TopLeft.Y + resolved.BottomLeft.Y)),
            Ratio(height, resolved.TopRight.Y + resolved.BottomRight.Y));
        if (scale < 1f)
        {
            resolved = resolved.Scaled(scale);
        }

        return resolved;
    }

    public bool IsZero() =>
        TopLeft.X.IsZero()
        && TopLeft.Y.IsZero()
        && TopRight.X.IsZero()
        && TopRight.Y.IsZero()
        && BottomRight.X.IsZero()
        && BottomRight.Y.IsZero()
        && BottomLeft.X.IsZero()
        && BottomLeft.Y.IsZero();

    private static (float X, float Y) ResolveCorner(CornerRadius radius, float width, float height)
    {
        float x = radius.X.Resolve(width);
        float y = radius.Y.Resolve(height);
        // CSS treats a corner as square when either axis has a zero radius.
        return x <= float.Epsilon || y <= float.Epsilon ? (0f, 0f) : (x, y);
    }
}

/// <summary>
/// Final used radii in CSS pixels after percentage resolution and overlap scaling.
/// </summary>
public readonly record struct ResolvedBorderRadii(
    (float X, float Y) TopLeft,
    (float X, float Y) TopRight,
    (float X, float Y) BottomRight,
    (float X, float Y) BottomLeft)
{
    public ResolvedBorderRadii Scaled(float scale)
    {
        static (float X, float Y) ScaleCorner((float X, float Y) corner, float scale) =>
            (corner.X * scale, corner.Y * scale);

        return new ResolvedBorderRadii(
            ScaleCorner(TopLeft, scale),
            ScaleCorner(TopRight, scale),
            ScaleCorner(BottomRight, scale),
            ScaleCorner(BottomLeft, scale));
    }

    public bool IsZero() =>
        TopLeft == (0f, 0f)
        && TopRight == (0f, 0f)
        && BottomRight == (0f, 0f)
        && BottomLeft == (0f, 0f);

    public bool IsUniform() =>
        TopLeft == TopRight && TopRight == BottomRight && BottomRight == BottomLeft;

    /// <summary>Inner radii after removing the used border widths on each axis.</summary>
    public ResolvedBorderRadii Inset(Sides<float> widths) => new(
        ShrinkCorner(TopLeft, widths.Left, widths.Top),
        ShrinkCorner(TopRight, widths.Right, widths.Top),
        ShrinkCorner(BottomRight, widths.Right, widths.Bottom),
        ShrinkCorner(BottomLeft, widths.Left, widths.Bottom));

    /// <summary>Outer radii for an outline or shadow expanded from the border edge.</summary>
    public ResolvedBorderRadii Outset(Sides<float> amounts) => new(
        GrowCorner(TopLeft, amounts.Left, amounts.Top),
        GrowCorner(TopRight, amounts.Right, amounts.Top),
        GrowCorner(BottomRight, amounts.Right, amounts.Bottom),
        GrowCorner(BottomLeft, amounts.Left, amounts.Bottom));

    private static (float X, float Y) ShrinkCorner(
        (float X, float Y) corner,
        float horizontal,
        float vertical)
    {
        float x = F32.Max(corner.X - horizontal, 0f);
        float y = F32.Max(corner.Y - vertical, 0f);
        return x <= float.Epsilon || y <= float.Epsilon ? (0f, 0f) : (x, y);
    }

    private static (float X, float Y) GrowCorner(
        (float X, float Y) corner,
        float horizontal,
        float vertical)
    {
        // Expanding a square corner must keep it square. Only an existing curve receives the
        // extra outline/shadow distance on each axis.
        if (corner.X <= float.Epsilon || corner.Y <= float.Epsilon)
        {
            return (0f, 0f);
        }

        float x = F32.Max(corner.X + horizontal, 0f);
        float y = F32.Max(corner.Y + vertical, 0f);
        return x <= float.Epsilon || y <= float.Epsilon ? (0f, 0f) : (x, y);
    }
}

/// <summary>Four independent specified border widths, styles, and colors.</summary>
/// <remarks>
/// A <c>null</c> color means CSS <c>currentcolor</c>, not transparent or absent. Rust's
/// <c>Default</c> seeds every specified width to <see cref="BorderSides.MediumBorderWidth"/>,
/// so <c>default(BorderModel)</c> is NOT the CSS initial value; use <see cref="Default"/>.
/// </remarks>
public readonly record struct BorderModel(
    Sides<float> SpecifiedWidths,
    Sides<BorderStyle> Styles,
    Sides<RgbaColor?> Colors,
    BorderRadii Radii)
{
    /// <summary>The CSS initial border state: <c>medium none currentcolor</c> on all sides.</summary>
    public static readonly BorderModel Default = new(
        Sides<float>.All(BorderSides.MediumBorderWidth),
        Sides<BorderStyle>.All(BorderStyle.None),
        Sides<RgbaColor?>.All(null),
        default);

    /// <summary>Used widths after applying CSS Backgrounds 3's <c>none</c>/<c>hidden</c> rule.</summary>
    public Sides<float> UsedWidths() => new(
        BorderSides.UsedWidth(SpecifiedWidths.Top, Styles.Top),
        BorderSides.UsedWidth(SpecifiedWidths.Right, Styles.Right),
        BorderSides.UsedWidth(SpecifiedWidths.Bottom, Styles.Bottom),
        BorderSides.UsedWidth(SpecifiedWidths.Left, Styles.Left));
}

/// <summary>
/// Uniform outline state. Outlines paint outside the border edge and never participate in
/// layout geometry.
/// </summary>
/// <remarks>
/// Rust's <c>Default</c> seeds the specified width to
/// <see cref="BorderSides.MediumBorderWidth"/>, so <c>default(OutlineModel)</c> is NOT the CSS
/// initial value; use <see cref="Default"/>.
/// </remarks>
public readonly record struct OutlineModel(
    float SpecifiedWidth,
    BorderStyle Style,
    RgbaColor? Color,
    float Offset)
{
    /// <summary>The CSS initial outline state.</summary>
    public static readonly OutlineModel Default =
        new(BorderSides.MediumBorderWidth, BorderStyle.None, null, 0f);

    public float UsedWidth() => BorderSides.UsedWidth(SpecifiedWidth, Style);
}

File: .reference/obscura/crates/obscura-render/src/border.rs
//! Typed CSS border and outline state shared by style, layout, and paint.
//!
//! CSS keeps the specified border width even while `none` or `hidden` makes
//! the *used* width zero. Keeping those two values distinct is what lets a
//! later `border-style: solid` restore the previously specified width without
//! corrupting box geometry in the meantime.

/// CSS's initial `medium` border and outline width in CSS pixels.
pub const MEDIUM_BORDER_WIDTH: f32 = 3.0;

/// Four physical sides in CSS top-right-bottom-left order.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Sides<T> {
    pub top: T,
    pub right: T,
    pub bottom: T,
    pub left: T,
}

impl<T: Copy> Sides<T> {
    pub const fn all(value: T) -> Self {
        Self {
            top: value,
            right: value,
            bottom: value,
            left: value,
        }
    }

    pub fn map<U>(self, mut f: impl FnMut(T) -> U) -> Sides<U> {
        Sides {
            top: f(self.top),
            right: f(self.right),
            bottom: f(self.bottom),
            left: f(self.left),
        }
    }

    pub const fn as_array(self) -> [T; 4] {
        [self.top, self.right, self.bottom, self.left]
    }
}

impl<T: Default> Default for Sides<T> {
    fn default() -> Self {
        Self {
            top: T::default(),
            right: T::default(),
            bottom: T::default(),
            left: T::default(),
        }
    }
}

/// Expand a CSS 1--4 value list in top-right-bottom-left order.
pub fn expand_sides<T: Copy>(values: &[T]) -> Option<Sides<T>> {
    match values {
        [all] => Some(Sides::all(*all)),
        [vertical, horizontal] => Some(Sides {
            top: *vertical,
            right: *horizontal,
            bottom: *vertical,
            left: *horizontal,
        }),
        [top, horizontal, bottom] => Some(Sides {
            top: *top,
            right: *horizontal,
            bottom: *bottom,
            left: *horizontal,
        }),
        [top, right, bottom, left] => Some(Sides {
            top: *top,
            right: *right,
            bottom: *bottom,
            left: *left,
        }),
        _ => None,
    }
}

/// The CSS border-line styles represented by the renderer.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum BorderStyle {
    #[default]
    None,
    Hidden,
    Dotted,
    Dashed,
    Solid,
    Double,
    Groove,
    Ridge,
    Inset,
    Outset,
    /// `outline-style:auto`; border sides themselves never use this value.
    Auto,
}

impl BorderStyle {
    /// Whether this style contributes a used width and visible border paint.
    pub const fn is_visible(self) -> bool {
        !matches!(self, Self::None | Self::Hidden)
    }

    pub const fn css_name(self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Hidden => "hidden",
            Self::Dotted => "dotted",
            Self::Dashed => "dashed",
            Self::Solid => "solid",
            Self::Double => "double",
            Self::Groove => "groove",
            Self::Ridge => "ridge",
            Self::Inset => "inset",
            Self::Outset => "outset",
            Self::Auto => "auto",
        }
    }
}

/// Four independent specified border widths, styles, and colors.
///
/// A `None` color means CSS `currentcolor`, not transparent or absent.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BorderModel {
    pub specified_widths: Sides<f32>,
    pub styles: Sides<BorderStyle>,
    pub colors: Sides<Option<[u8; 4]>>,
    pub radii: BorderRadii,
}

impl Default for BorderModel {
    fn default() -> Self {
        Self {
            specified_widths: Sides::all(MEDIUM_BORDER_WIDTH),
            styles: Sides::all(BorderStyle::None),
            colors: Sides::all(None),
            radii: BorderRadii::default(),
        }
    }
}

impl BorderModel {
    /// Used widths after applying CSS Backgrounds 3's `none`/`hidden` rule.
    pub fn used_widths(self) -> Sides<f32> {
        Sides {
            top: used_width(self.specified_widths.top, self.styles.top),
            right: used_width(self.specified_widths.right, self.styles.right),
            bottom: used_width(self.specified_widths.bottom, self.styles.bottom),
            left: used_width(self.specified_widths.left, self.styles.left),
        }
    }
}

fn used_width(specified: f32, style: BorderStyle) -> f32 {
    if style.is_visible() {
        specified.max(0.0)
    } else {
        0.0
    }
}

/// Uniform outline state. Outlines paint outside the border edge and never
/// participate in layout geometry.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct OutlineModel {
    pub specified_width: f32,
    pub style: BorderStyle,
    /// `None` is CSS `currentcolor`.
    pub color: Option<[u8; 4]>,
    pub offset: f32,
}

impl Default for OutlineModel {
    fn default() -> Self {
        Self {
            specified_width: MEDIUM_BORDER_WIDTH,
            style: BorderStyle::None,
            color: None,
            offset: 0.0,
        }
    }
}

impl OutlineModel {
    pub fn used_width(self) -> f32 {
        used_width(self.specified_width, self.style)
    }
}

/// One CSS `<length-percentage>` component of a corner radius.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct RadiusValue {
    pub length: f32,
    /// Fraction of the corresponding border-box axis (`0.5` is `50%`).
    pub percentage: f32,
}

impl RadiusValue {
    pub const fn pixels(length: f32) -> Self {
        Self {
            length,
            percentage: 0.0,
        }
    }

    pub const fn percentage(percentage: f32) -> Self {
        Self {
            length: 0.0,
            percentage,
        }
    }

    pub fn resolve(self, axis: f32) -> f32 {
        (self.length + self.percentage * axis).max(0.0)
    }

    pub fn is_zero(self) -> bool {
        self.length == 0.0 && self.percentage == 0.0
    }
}

/// Horizontal and vertical radii for one corner.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct CornerRadius {
    pub x: RadiusValue,
    pub y: RadiusValue,
}

impl CornerRadius {
    pub const fn circular(value: RadiusValue) -> Self {
        Self { x: value, y: value }
    }
}

/// Four elliptical radii in clockwise order from the top-left corner.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct BorderRadii {
    pub top_left: CornerRadius,
    pub top_right: CornerRadius,
    pub bottom_right: CornerRadius,
    pub bottom_left: CornerRadius,
}

impl BorderRadii {
    pub const fn all(radius: CornerRadius) -> Self {
        Self {
            top_left: radius,
            top_right: radius,
            bottom_right: radius,
            bottom_left: radius,
        }
    }

    /// Resolve percentages and apply CSS Backgrounds 3's corner-overlap rule.
    /// One common scale factor preserves every ellipse's aspect ratio.
    pub fn resolve(self, width: f32, height: f32) -> ResolvedBorderRadii {
        let mut resolved = ResolvedBorderRadii {
            top_left: resolve_corner(self.top_left, width, height),
            top_right: resolve_corner(self.top_right, width, height),
            bottom_right: resolve_corner(self.bottom_right, width, height),
            bottom_left: resolve_corner(self.bottom_left, width, height),
        };
        let ratio = |available: f32, requested: f32| {
            if requested > 0.0 {
                available.max(0.0) / requested
            } else {
                1.0
            }
        };
        let scale = 1.0f32
            .min(ratio(width, resolved.top_left.0 + resolved.top_right.0))
            .min(ratio(width, resolved.bottom_left.0 + resolved.bottom_right.0))
            .min(ratio(height, resolved.top_left.1 + resolved.bottom_left.1))
            .min(ratio(height, resolved.top_right.1 + resolved.bottom_right.1));
        if scale < 1.0 {
            resolved = resolved.scaled(scale);
        }
        resolved
    }

    pub fn is_zero(self) -> bool {
        self.top_left.x.is_zero()
            && self.top_left.y.is_zero()
            && self.top_right.x.is_zero()
            && self.top_right.y.is_zero()
            && self.bottom_right.x.is_zero()
            && self.bottom_right.y.is_zero()
            && self.bottom_left.x.is_zero()
            && self.bottom_left.y.is_zero()
    }
}

fn resolve_corner(radius: CornerRadius, width: f32, height: f32) -> (f32, f32) {
    let x = radius.x.resolve(width);
    let y = radius.y.resolve(height);
    // CSS treats a corner as square when either axis has a zero radius.
    if x <= f32::EPSILON || y <= f32::EPSILON {
        (0.0, 0.0)
    } else {
        (x, y)
    }
}

/// Final used radii in CSS pixels after percentage resolution and overlap
/// scaling.
#[derive(Debug, Clone, Copy, Default, PartialEq)]
pub struct ResolvedBorderRadii {
    pub top_left: (f32, f32),
    pub top_right: (f32, f32),
    pub bottom_right: (f32, f32),
    pub bottom_left: (f32, f32),
}

impl ResolvedBorderRadii {
    pub fn scaled(self, scale: f32) -> Self {
        let scale_corner = |(x, y): (f32, f32)| (x * scale, y * scale);
        Self {
            top_left: scale_corner(self.top_left),
            top_right: scale_corner(self.top_right),
            bottom_right: scale_corner(self.bottom_right),
            bottom_left: scale_corner(self.bottom_left),
        }
    }

    pub fn is_zero(self) -> bool {
        self.top_left == (0.0, 0.0)
            && self.top_right == (0.0, 0.0)
            && self.bottom_right == (0.0, 0.0)
            && self.bottom_left == (0.0, 0.0)
    }

    pub fn is_uniform(self) -> bool {
        self.top_left == self.top_right
            && self.top_right == self.bottom_right
            && self.bottom_right == self.bottom_left
    }

    /// Inner radii after removing the used border widths on each axis.
    pub fn inset(self, widths: Sides<f32>) -> Self {
        Self {
            top_left: shrink_corner(self.top_left, widths.left, widths.top),
            top_right: shrink_corner(self.top_right, widths.right, widths.top),
            bottom_right: shrink_corner(
                self.bottom_right,
                widths.right,
                widths.bottom,
            ),
            bottom_left: shrink_corner(self.bottom_left, widths.left, widths.bottom),
        }
    }

    /// Outer radii for an outline or shadow expanded from the border edge.
    pub fn outset(self, amounts: Sides<f32>) -> Self {
        Self {
            top_left: grow_corner(self.top_left, amounts.left, amounts.top),
            top_right: grow_corner(self.top_right, amounts.right, amounts.top),
            bottom_right: grow_corner(
                self.bottom_right,
                amounts.right,
                amounts.bottom,
            ),
            bottom_left: grow_corner(self.bottom_left, amounts.left, amounts.bottom),
        }
    }
}

fn shrink_corner((x, y): (f32, f32), horizontal: f32, vertical: f32) -> (f32, f32) {
    let x = (x - horizontal).max(0.0);
    let y = (y - vertical).max(0.0);
    if x <= f32::EPSILON || y <= f32::EPSILON {
        (0.0, 0.0)
    } else {
        (x, y)
    }
}

fn grow_corner((x, y): (f32, f32), horizontal: f32, vertical: f32) -> (f32, f32) {
    // Expanding a square corner must keep it square. Only an existing curve
    // receives the extra outline/shadow distance on each axis.
    if x <= f32::EPSILON || y <= f32::EPSILON {
        return (0.0, 0.0);
    }
    let x = (x + horizontal).max(0.0);
    let y = (y + vertical).max(0.0);
    if x <= f32::EPSILON || y <= f32::EPSILON {
        (0.0, 0.0)
    } else {
        (x, y)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn side_expansion_follows_css_trbl_rules() {
        assert_eq!(expand_sides(&[1]), Some(Sides::all(1)));
        assert_eq!(
            expand_sides(&[1, 2, 3]),
            Some(Sides {
                top: 1,
                right: 2,
                bottom: 3,
                left: 2,
            })
        );
        assert_eq!(expand_sides::<i32>(&[]), None);
        assert_eq!(expand_sides(&[1, 2, 3, 4, 5]), None);
    }

    #[test]
    fn hidden_style_zeroes_only_the_used_width() {
        let mut border = BorderModel::default();
        border.specified_widths = Sides::all(10.0);
        border.styles = Sides::all(BorderStyle::None);
        assert_eq!(border.used_widths(), Sides::all(0.0));
        border.styles.left = BorderStyle::Solid;
        assert_eq!(border.used_widths().left, 10.0);
        assert_eq!(border.specified_widths.left, 10.0);
    }

    #[test]
    fn elliptical_radii_use_one_overlap_scale() {
        let radii = BorderRadii {
            top_left: CornerRadius {
                x: RadiusValue::pixels(80.0),
                y: RadiusValue::pixels(50.0),
            },
            top_right: CornerRadius {
                x: RadiusValue::pixels(60.0),
                y: RadiusValue::pixels(40.0),
            },
            bottom_right: CornerRadius {
                x: RadiusValue::pixels(40.0),
                y: RadiusValue::pixels(30.0),
            },
            bottom_left: CornerRadius {
                x: RadiusValue::pixels(20.0),
                y: RadiusValue::pixels(10.0),
            },
        };
        let resolved = radii.resolve(100.0, 50.0);
        let scale = 5.0 / 7.0;
        assert!((resolved.top_left.0 - 80.0 * scale).abs() < 0.001);
        assert!((resolved.top_left.1 - 50.0 * scale).abs() < 0.001);
        assert!((resolved.bottom_right.0 - 40.0 * scale).abs() < 0.001);
        assert!((resolved.bottom_left.1 - 10.0 * scale).abs() < 0.001);
    }
}
